File: python/day4.py
#!/usr/bin/env python3

import itertools
from pathlib import Path
import typing as t

import numpy
from numpy.typing import NDArray
import scipy.ndimage


def load_array(input_path: t.Union[str, Path]) -> NDArray[numpy.bool_]:
    arr = []
    with open(input_path, 'rb') as f:
        for line in f:
            arr.append(numpy.frombuffer(line.strip(), dtype=numpy.uint8))

    arr = numpy.stack(arr, axis=0)
    return arr == b'@'[0]

def print_array(arr: NDArray[numpy.bool_]):
    print(numpy.concatenate([
        numpy.where(arr, b'@'[0], b'.'[0]),
        numpy.full((arr.shape[0], 1), b'\n'[0])
    ], axis=1).tobytes().decode('ascii'), end='\n')

def process(input_path: t.Union[str, Path], verbose: bool = False):
    arr = load_array(input_path)
    weights = numpy.ones((3,), numpy.uint8)
    n_total = numpy.sum(arr, dtype=numpy.uint64)

    if verbose:
        print("Initial state:")
        print_array(arr)

    for i in itertools.count(1):
        convolved = arr.astype(numpy.uint8)
        for axis in (0, 1):
            convolved = scipy.ndimage.convolve1d(convolved, weights, axis=axis, mode='constant')

        available = arr & (convolved < 5)  # 4 neighbors + self
        n_available = numpy.sum(available)
        print(f"Step {i}, {n_available:3} box(es) available")
        if not n_available:
            break
        if verbose:
            print_array(arr)

        arr ^= available

    print(f"Finished in {i} step(s), final state:")
    print_array(arr)
    n_final = numpy.sum(arr, dtype=numpy.uint64)
    print(f"{n_total} -> {n_final} boxes (removed {n_total - n_final})")


def main():
    input_folder = Path(__file__).absolute().parent.parent / 'input'
    input_path = input_folder / 'day4.txt'

    process(input_path)

if __name__ == '__main__':
    main()

File: rust/src/day4.rs
use std::{fs::File, io::{BufRead, BufReader}, path::Path};

use ndarray::{Array1, Array2, ArrayBase, Axis, Data, Dim, stack};
use ndarray_conv::{ConvExt, ConvMode, PaddingMode};

use super::{input_dir, verbosity};


fn load_array<P: AsRef<Path>>(path: P) -> Result<Array2<u8>, String> {
    let file = File::open(path).map_err(|e| format!("Failed to open input file: {e}"))?;
    let mut rows = Vec::new();

    for line in BufReader::new(file).lines() {
        let line = line.map_err(|e| format!("Error reading file: {e}"))?;
        rows.push(Array1::from_iter(line.bytes().map(|c| (c == b'@') as u8)))
    }

    let views: Vec<_> = rows.iter().map(|arr| arr.view()).collect();
    stack(Axis(0), &views).map_err(|_| format!("Mismatched shapes in rows"))
}

fn print_array<'a, S>(arr: ArrayBase<S, Dim<[usize; 2]>>)
where S: Data<Elem = u8> + 'a
{
    for line in arr.outer_iter() {
        let line: String = line.iter().map(|c| if *c > 0 { '@' } else { '.' }).collect();
        println!("{}", line)
    }
    println!("")
}

#[allow(unused)]
fn print_counts<'a, S>(arr: ArrayBase<S, Dim<[usize; 2]>>)
where S: Data<Elem = u8> + 'a
{
    for line in arr.outer_iter() {
        let line: String = line.iter().map(|c| {
            assert!(*c < 10);
            (c + 48) as char
        }).collect();
        println!("{}", line)
    }
    println!("")
}

pub fn process(mut arr: Array2<u8>, verbosity: u8) -> u64 {
    let n_start: u64 = arr.iter().map(|c| *c as u64).sum();
    let mut n = n_start;

    if verbosity > 0 {
        println!("Initial state:");
        print_array(arr.view());
    }

    // kernel is separable, use 2 1D convolutions
    let kernel1: Array2<u8> = Array2::ones((3, 1));
    let kernel2: Array2<u8> = Array2::ones((1, 3));

    let mut i = 0;

    loop {
        let convolved = ConvExt::conv(
            &ConvExt::conv(&arr, &kernel1, ConvMode::Same, PaddingMode::Zeros).unwrap(),
            &kernel2, ConvMode::Same, PaddingMode::Zeros
        ).unwrap();
        if verbosity > 1 {
            print_counts(convolved.view());
        }

        // 4 neighbors + self
        let mut available = convolved.mapv(|c| if c < 5 { 1u8 } else { 0u8 });
        available &= &arr;
        let n_available: u64 = available.iter().map(|c| *c as u64).sum();
        if n_available == 0 { break; }

        i += 1;
        println!("Step {i}, removed {n_available:3} box(es)");
        arr ^= &available;
        n -= n_available;
        if verbosity > 0 { print_array(arr.view()); }
    }

    println!("Finished in {i} step(s), final state:");
    print_array(arr);
    println!("{n_start} -> {n} boxes (removed {})", n_start - n);
    n_start - n
}


pub fn run(test: bool) -> Result<(), String> {
    let verbosity = verbosity();

    let mut input_path = input_dir().to_owned();
    input_path.push(if test { "day4_test.txt" } else { "day4.txt" });

    let arr = load_array(&input_path)?;
    process(arr, verbosity);

    Ok(())
}